### PR TITLE
Add disk spilling support for hash join build side

### PR DESCRIPTION
**Pull request by xiaoxmeng**
Add disk spilling support for hash join. It mainly works as follows:
(1) task will create spill operator group before driver creation;
(2) hash build operator add into spill operator group and hash join bridge
on construction;
(3) hash build operator create spiller on construction and we could
optimize this to create on demand;
(4) task starts spill operator group;
(5) hash build operator add to reserve memory for new input if disk
spilling is enabled. If it fails to reserve memory and it will send request
to spill operator group to run group spill;
(6) hash build operators wait for group spill to run which pickup the
spill partitions and flush out all the buffered rows into spill files;
(7) hash build operator split the new input by writing rows to spill file
directly if the corresponding partitions are spilling, and then update
row selection vector to continue the regular path with non-spilling rows;
(8) after all the operators finishes processing data, the last build operator
will collect spilled partitions from all the builders and set it into the hash
bridge along with the built table;
(9) if there is no more spill data, then build operator finish, otherwise wait
for spill input from the prober side, and the prober side will set it after it
finishes the processing (in followup);
(10) once get the spill input from the prober side, the build operator will
reset the spiller and create spill input reader to process spilled data which
pretty much follow the previous steps as process data from the build source.
The difference is that the input is processed in loop until reaches the end
of the spill data or run out of memory that waits for recursive group spill.

This PR adds simple test with empty probe to verify that the spill has been
triggered from the build side. The followup with prober side change will
turn on spilling test coverage on all the existing test cases in HashJoinTest
plus some dedicated test cases.

### CODE REVIEW

File: velox/exec/HashBuild.h [hunk 1]
```diff
@@ -15,38 +15,106 @@
...
-// and probe Operator instances concerned. Corresponds to the Presto concept of
-// the same name.
+/// Hands over a hash table from a multi-threaded build pipeline to a
+/// multi-threaded probe pipeline. This is owned by shared_ptr by all the build
+/// and probe Operator instances concerned. Corresponds to the Presto concept of
+/// the same name.
 class HashJoinBridge : public JoinBridge {
  public:
-  void setHashTable(std::unique_ptr<BaseHashTable> table);
+  HashJoinBridge() = default;
```
Comment (mbasmanova) [inline velox/exec/HashBuild.h:36]: Why is this needed?

File: velox/exec/HashBuild.h [hunk 2]
```diff
@@ -15,38 +15,106 @@
...
+        std::optional<SpillPartitionId> _restoredPartitionId,
+        SpillPartitionIdSet _spillPartitionIds)
+        : antiJoinHasNullKeys(false),
+          table(std::move(_table)),
+          restoredPartitionId(std::move(_restoredPartitionId)),
+          spillPartitionIds(std::move(_spillPartitionIds)) {}
+
+    HashBuildResult() : antiJoinHasNullKeys(true) {}
+
+    bool antiJoinHasNullKeys = false;
```
Comment (mbasmanova) [inline velox/exec/HashBuild.h:71]: This can be const, right?

File: velox/exec/HashBuild.cpp [hunk 3]
```diff
@@ -51,24 +120,97 @@ void HashJoinBridge::setAntiJoinHasNullKeys() {
...
+
+std::optional<HashJoinBridge::SpillInput> HashJoinBridge::spillInputOrFuture(
+    ContinueFuture* future) {
+  std::lock_guard<std::mutex> l(mutex_);
+  VELOX_DCHECK(init_);
+  VELOX_CHECK(!cancelled_, "Getting spill input after join is aborted");
+  VELOX_DCHECK(
+      !restoringSpillPartitionId_.has_value() || !buildResult_.has_value());
+
+  // TODO: enable the following code once the probe side has spilling support.
```
Comment (mbasmanova) [inline velox/exec/HashBuild.cpp:173]: Let's remove commented out code.

File: velox/exec/HashBuild.h [hunk 4]
```diff
@@ -15,38 +15,106 @@
...
+
+    SpillInput(std::unique_ptr<SpillPartition> spillPartition)
+        : noMoreInput(false), spillPartition(std::move(spillPartition)) {}
+
+    bool noMoreInput{false};
+    std::unique_ptr<SpillPartition> spillPartition;
   };
 
-  std::optional<HashBuildResult> tableOrFuture(ContinueFuture* future);
+  std::optional<SpillInput> spillInputOrFuture(
```
Comment (mbasmanova) [inline velox/exec/HashBuild.h:98]: Would you document this method? Who is calling it? What is it used for?

**Review by mbasmanova — COMMENTED** [review]
@xiaoxmeng This is a very large PR. As discussed offline, let's split it up into smaller easier to review pieces. I haven't finished reading HashBuild.* yet. Will continue tomorrow.

**Comment by oerling** [comment]
Here's a few comments.

Generally looks fine.

 

Thanks

Orri

 

 

HashBuild.cpp / .h

 

joinBride_ -> join

 

+    /// The state tht waits for the hash tables to be merged together.

 

 

tht -> that

Bridge_

 

 

+  // the build source data type if the input is read from spill data during

+  // restoring.

+  RowTypePtr inputType() const;

+

 

 

Q: Different  in what way?

 

 

***

 

The term split is confusing. Maybe could call it fragment instead. It has a role that is like a split for scan but it does not go via the same channels.

 

 

***

 

  // Invoked by 'runSpill' to sum up the spill targets from all the operators in

  // 'numRows' and 'numBytes'.

 

 

 

Q: Maybe have comments as active sentences like 'Returns row and byte count for spilled data in 'numRows' and 'numBytes'. Resets the counters. Used by ... for ....

File: velox/exec/Task.cpp [hunk 5]
```diff
@@ -1843,4 +1852,19 @@ std::string Task::getErrorMsgOnMemCapExceeded(
...
+SpillOperatorGroup* Task::getSpillOperatorGroupLocked(
+    uint32_t splitGroupId,
+    const core::PlanNodeId& planNodeId) {
+  auto& groups = splitGroupStates_[splitGroupId].spillOperatorGroups;
+  auto it = groups.find(planNodeId);
+  VELOX_CHECK(it != groups.end(), "Split group is not set {}", splitGroupId);
+  SpillOperatorGroup* group = it->second.get();
+  VELOX_CHECK_NOT_NULL(
+      group,
+      "Spill group for plan node ID {} is not set in split group {}",
```
Comment (mbasmanova) [inline velox/exec/Task.cpp:1865]: "spill group" and "split group" in the same sentence are hard to distinguish.

File: velox/exec/Driver.h [hunk 6]
```diff
@@ -124,8 +124,14 @@ enum class BlockingReason {
   kWaitForSplit,
   kWaitForExchange,
   kWaitForJoinBuild,
+  /// Build operator is blocked waiting for the probe operators to finish
```
Comment (mbasmanova) [inline velox/exec/Driver.h:127]: Nice comments.

File: velox/exec/HashBuild.h [hunk 7]
```diff
@@ -59,19 +81,153 @@ class HashBuild final : public Operator {
...
+  // the memory reservation succeeds. If the memory reservation fails, the
+  // function will trigger a group spill which needs coordination among the
+  // other build drivers in the same group. The function returns true if the
+  // group spill has been inline executed which could happen if there is only
+  // one driver in the group, or it happens that all the other drivers have
+  // also requested group spill and this driver is the last one to reach the
+  // group spill barrier. Otherwise, the function returns false to wait for the
+  // group spill to run. The operator will transition to 'kWaitForSpill' state
+  // accordingly.
+  bool ensureInputFits(RowVectorPtr& input);
```
Comment (mbasmanova) [inline velox/exec/HashBuild.h:149]: Should the argument be passed as const &?

File: velox/exec/HashBuild.cpp [hunk 8]
```diff
@@ -178,12 +278,315 @@ void HashBuild::addInput(RowVectorPtr input) {
...
+    const auto partition = spillPartitions_[row];
+    if (FOLLY_UNLIKELY(!activeRows_.isValid(row))) {
+      continue;
+    }
+    if (!spiller_->isSpilled(partition)) {
+      continue;
+    }
+    activeRows_.setValid(row, false);
+    ++numSpillInputs;
+    spillInputIndicesBuffers_[partition]
```
Comment (mbasmanova) [inline velox/exec/HashBuild.cpp:391]: Would it make sense to cache raw pointers to the buffers to avoid the cast for each row?

File: velox/exec/HashBuild.cpp [hunk 9]
```diff
@@ -178,12 +278,315 @@ void HashBuild::addInput(RowVectorPtr input) {
...
+       ++partition) {
+    const int numSpillInputs = numSpillInputs_[partition];
+    if (numSpillInputs == 0) {
+      continue;
+    }
+    VELOX_CHECK(spiller_->isSpilled(partition));
+    spillPartition(
+        partition, numSpillInputs, spillInputIndicesBuffers_[partition], input);
+  }
+  activeRows_.updateBounds();
```
Comment (mbasmanova) [inline velox/exec/HashBuild.cpp:408]: perhaps, move this up right after the loop which updates activeRows_

File: velox/exec/HashBuild.cpp [hunk 10]
```diff
@@ -178,12 +278,315 @@ void HashBuild::addInput(RowVectorPtr input) {
...
+}
+
+void HashBuild::prepareInputIndicesBuffers(
+    vector_size_t numInput,
+    const SpillPartitionNumSet& spillPartitions) {
+  for (const auto& partition : spillPartitions) {
+    if (spillInputIndicesBuffers_[partition] == nullptr) {
+      spillInputIndicesBuffers_[partition] = allocateIndices(numInput, pool());
+    } else {
+      AlignedBuffer::reallocate<vector_size_t>(
```
Comment (mbasmanova) [inline velox/exec/HashBuild.cpp:418]: reallocate copies the old data in case it needs to allocate new buffer; since we don't need the old data, it might be more efficient to just check if buffer is large enough and call allocate if not.

File: velox/exec/HashBuild.cpp [hunk 11]
```diff
@@ -178,12 +278,315 @@ void HashBuild::addInput(RowVectorPtr input) {
...
+          &spillInputIndicesBuffers_[partition], numInput);
+    }
+  }
+  std::fill(numSpillInputs_.begin(), numSpillInputs_.end(), 0);
+}
+
+void HashBuild::computeSpillPartitions(const RowVectorPtr& input) {
+  if (hashes_.size() < activeRows_.size()) {
```
Comment (mbasmanova) [inline velox/exec/HashBuild.cpp:426]: nit: we only need activeRows_.end() entries

File: velox/exec/HashBuild.cpp [hunk 12]
```diff
@@ -178,12 +278,315 @@ void HashBuild::addInput(RowVectorPtr input) {
...
+    vector_size_t size,
+    BufferPtr indices,
+    const RowVectorPtr& input) {
+  VELOX_DCHECK(spillEnabled());
+
+  if (isInputFromSpill()) {
+    spiller_->spill(partition, wrap(size, std::move(indices), input));
+  } else {
+    int spillChannel = 0;
+    for (const auto& channel : keyChannels_) {
```
Comment (mbasmanova) [inline velox/exec/HashBuild.cpp:456]: Do we need to run these loops for each partition?

File: velox/exec/HashBuild.cpp [hunk 13]
```diff
@@ -178,12 +278,315 @@ void HashBuild::addInput(RowVectorPtr input) {
...
+    return false;
+  }
+
+  input_ = std::move(input);
+  if (spillGroup_->waitSpill(*this, future_)) {
+    VELOX_CHECK(future_.valid());
+    setState(State::kWaitForSpill);
+    return true;
+  }
+  input = std::move(input_);
```
Comment (mbasmanova) [inline velox/exec/HashBuild.cpp:498]: Why do we need to move input to input_, then back? Could we move input only if spillGroup_->waitSpill returns true?

**Review by mbasmanova — COMMENTED** [review]
@xiaoxmeng Read till L502 in HashBuild.cpp. Have some questions.

**Review by mbasmanova — APPROVED** [review]
@xiaoxmeng Looks great % some questions I posted earlier.

File: velox/exec/HashBuild.cpp (hunk 10 above)
Comment (xiaoxmeng) [reply velox/exec/HashBuild.cpp]: Good point!

File: velox/exec/HashBuild.cpp (hunk 12 above)
Comment (xiaoxmeng) [reply velox/exec/HashBuild.cpp]: Good point. Thanks!

File: velox/exec/HashBuild.cpp (hunk 8 above)
Comment (xiaoxmeng) [reply velox/exec/HashBuild.cpp]: Yeah, good point. Updated!

File: velox/exec/HashBuild.cpp (hunk 13 above)
Comment (xiaoxmeng) [reply velox/exec/HashBuild.cpp]: Yes. Thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@mbasmanova Thanks for the quick review. I have addressed most of your comments. Could you please take a second look?

**Comment by markjin1990** [comment]
@xiaoxmeng @mbasmanova we noticed at ByteDance that this commit caused significant performance regression on TPCH (SF1000) performance benchmarks, when spilling is certainly disabled on our deployment. With this commit, Q8 takes >3mins, but if we revert this commit, Q8 takes around 1.5mins. Similar performance regression also happens to queries like Q3, Q9, Q21 with this commit. The performance regression is mostly caused by a slower HashJoin, which is quite confusing given that ideally this commit should not make an impact on hash join performance if spilling is disabled. 
So far, we have not identified the actual root cause for issue we observed in this commit. No sure if you have encountered a similar performance issue at Meta?

PS: 
One thing we noticed might be a mistake in this commit. However, this is not the root cause of the performance regression we observed. 
image

**Comment by mbasmanova** [comment]
@markjin1990 Thank you so much for reporting this performance regression. It would be very helpful if you could open a GitHub issue. We'll take a look.

**Comment by markjin1990** [comment]
> @markjin1990 Thank you so much for reporting this performance regression. It would be very helpful if you could open a GitHub issue. We'll take a look.

Just did. https://github.com/facebookincubator/velox/issues/2987 Thank you!

File: velox/exec/HashBuild.cpp [hunk 14]
```diff
@@ -178,12 +277,316 @@ void HashBuild::addInput(RowVectorPtr input) {
...
+  std::sort(
+      partitionIndices.begin(),
+      partitionIndices.end(),
+      [&](int32_t lhs, int32_t rhs) {
+        return spillableStats[lhs].numBytes > spillableStats[rhs].numBytes;
+      });
+  int64_t numRows = 0;
+  int64_t numBytes = 0;
+  for (auto partitionNum : partitionIndices) {
+    if (spillableStats[partitionNum].numBytes == 0) {
```
Comment (xudong963) [inline velox/exec/HashBuild.cpp:538]: Hi @xiaoxmeng , why directly break if one partition's bytes are zero? It's possible that the following partitions aren't empty, no?
